modules/partition-replicator/src/integrationTest/java/org/apache/ignite/internal/partition/replicator/ItBuildIndexTest.java

```diff
@@ -0,0 +1,155 @@
+
+        String zoneName = "test-zone";
+        createZone(node, zoneName, 1, 1);
+
+        String tableName = "TEST_TABLE";
+        createCustomTable(node, zoneName, tableName);
+        int tableId = TableTestUtils.getTableId(node.catalogManager, tableName, node.hybridClock.nowLong());
+
+        // Test node does not create IndexNodeFinishedRwTransactionsChecker, so the following code is needed to unblock index building.
+        // It's easier than creating a real service with all its dependencies.
+        node.clusterService.messagingService().addMessageHandler(
+                IndexMessageGroup.class,
+                (message, sender, correlationId) -> {
+                    if (message instanceof IsNodeFinishedRwTransactionsStartedBeforeRequest) {
+                        node.clusterService.messagingService().respond(
+                                sender,
+                                FACTORY.isNodeFinishedRwTransactionsStartedBeforeResponse().finished(true).build(),
+                                correlationId
+                        );
+                    }
+                });
+
+        // Firing this event is a workaround to trigger the index building process.
+        // I agree that this approach is far from a good one and looks like a hack,
+        // but it's the only way to unblock the index building process using our test node.
+        placementDriver.fireTestEvent(
+                PRIMARY_REPLICA_ELECTED,
+                new PrimaryReplicaEventParameters(
+                        85, // causalityToken,
+                        new TablePartitionId(tableId, 0), // replicationGroupId
+                        node.clusterService.topologyService().localMember().id(), // leaseholderId,
+                        node.clusterService.topologyService().localMember().name(), // lease holder name
+                        node.hybridClock.now() // lease start time
+                ));
+
+        TableViewInternal tableViewInternal = node.tableManager.table(tableId);
+        KeyValueView<Tuple, Tuple> tableView = tableViewInternal.keyValueView();
+
+        node.transactions().runInTransaction(tx -> {
+            Tuple key = Tuple.create().set("KEY", 1L);
+            Tuple value = Tuple.create().set("VAL", 1).set("DOUBLEVAL", 1.0);
+            tableView.putAll(tx, Map.of(key, value));
+        });
+
+        // This async transaction is needed to update safe time, just because the idle safe time propagation is not implemented yet.
+        // https://issues.apache.org/jira/browse/IGNITE-22620
+        CompletableFuture.runAsync(() -> {
+            try {
+                Thread.sleep(1_000);
+                node.transactions().runInTransaction(tx -> {
+                    Tuple key = Tuple.create().set("KEY", 1L);
+                    Tuple value = Tuple.create().set("VAL", 1).set("DOUBLEVAL", 1.0);
+                    tableView.putAll(tx, Map.of(key, value));
```

Comment:
Why is second put needed for the same key?

Reply:
Idle safe-time propagation has not been implemented yet on the colocation track.

Reply:
How is second put related to idle safe time propagation? :)